engine/engine.go

```diff
@@ -237,16 +238,44 @@ func (e *EngineAPI) GetPayloadV3(payloadID engine.PayloadID) (*eth.ExecutionPayl
 
 	// TODO: handle time slot based block production
 	// for now assume block is sealed by this call
-	if err := e.builder.Build(&builder.Payload{
-		Transactions: e.currentPayloadAttributes.CosmosTxs,
-		GasLimit:     e.currentPayloadAttributes.GasLimit,
-		Timestamp:    e.currentPayloadAttributes.Timestamp,
-		NoTxPool:     e.currentPayloadAttributes.NoTxPool,
-	}); err != nil {
+	block, err := e.builder.Build(&builder.Payload{
+		InjectedTransactions: e.currentPayloadAttributes.CosmosTxs,
+		GasLimit:             e.currentPayloadAttributes.GasLimit,
+		Timestamp:            e.currentPayloadAttributes.Timestamp,
+		NoTxPool:             e.currentPayloadAttributes.NoTxPool,
+	})
+	if err != nil {
 		log.Panicf("failed to commit block: %v", err) // TODO error handling. An error here is potentially a big problem.
 	}
-	payloadEnvelope := e.currentPayloadAttributes.ToExecutionPayloadEnvelope(e.blockStore.HeadBlock().Hash())
 
+	txs, err := e.adapter.CosmosToEth(block.Txs)
+	if err != nil {
+		return nil, engine.GenericServerError.With(fmt.Errorf("convert cosmos txs to eth txs: %v", err))
+	}
+
+	txBytes := make([]hexutil.Bytes, len(txs))
+	for i, tx := range txs {
+		// return deposit transactions directly from the received payload attributes
+		if i < len(e.currentPayloadAttributes.Transactions) {
+			txBytes[i] = e.currentPayloadAttributes.Transactions[i]
+		} else {
+			txBytes[i] = tx.Data()
+		}
```

Comment:
```suggestion
		txBytes[i] = tx.Data()
```
Any reason we can do this?

Reply:
That was the prior attempt - doesn't work because of some mismatched data (types?).

See https://www.notion.so/nethermind/36d9017dfafc440e96a8616b269bf619?v=90d07fe7c34e40109dd229c7605d7ca1&p=f071819347f048afa1cfa6c9c86d4c06&pm=s for some logged details from e2e tests.

I expect this to get resolved as the adapters become better tested.

Reply:
Ah, I see what I did wrong. `tx.Data()` only returns the data field of a deposit tx; it doesn't marshal the tx itself.

This should work:

```go
		txBytes[i], err = tx.MarshalBinary()
		if err != nil {
			return nil, engine.GenericServerError.With(fmt.Errorf("marshal tx binary: %v", err))
		}
```

Reply:
Oh, great. I had used another marshalling function unsuccessfully (ie, it failed the tests) and maybe confused it with this function.

ccef2f1